File: android_webview/test/shell/memory_metrics_logger_utils.cc
// Copyright 2019 The Chromium Authors
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#include "android_webview/browser/metrics/memory_metrics_logger.h"
#include "base/synchronization/waitable_event.h"
#include "base/task/sequenced_task_runner.h"
#include "base/task/single_thread_task_runner.h"
#include "base/test/bind.h"
#include "base/test/test_timeouts.h"

// Must come after all headers that specialize FromJniType() / ToJniType().
#include "android_webview/test/webview_instrumentation_test_native_jni/MemoryMetricsLoggerUtils_jni.h"

namespace android_webview {

// static
static jboolean JNI_MemoryMetricsLoggerUtils_ForceRecordHistograms(
    JNIEnv* env) {
  // Note: Using a WaitableEvent instead of a RunLoop because there is no
  // sequenced context. Also can't use a TaskEnvironment because this function
  // is not running on the main thread.
  CHECK(!base::SequencedTaskRunner::HasCurrentDefault());
  CHECK(!base::SingleThreadTaskRunner::GetMainThreadDefault()
             ->BelongsToCurrentThread());

  auto* memory_metrics_logger =
      ::metrics::MemoryMetricsLogger::GetInstanceForTesting();
  if (!memory_metrics_logger) {
    return false;
  }

  TestTimeouts::Initialize();
  base::WaitableEvent waitable_event;
  bool result = false;
  memory_metrics_logger->ScheduleRecordForTesting(
      base::BindLambdaForTesting([&](bool success) {
        result = success;
        waitable_event.Signal();
      }));
  waitable_event.TimedWait(TestTimeouts::action_timeout());
  return result;
}

}  // namespace android_webview

DEFINE_JNI(MemoryMetricsLoggerUtils)

File: android_webview/test/shell/src/org/chromium/android_webview/test/util/MemoryMetricsLoggerUtils.java
// Copyright 2019 The Chromium Authors
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

package org.chromium.android_webview.test.util;

import org.jni_zero.JNINamespace;
import org.jni_zero.NativeMethods;

/** Utilities for force recording memory metrics. */
@JNINamespace("android_webview")
public class MemoryMetricsLoggerUtils {
    @NativeMethods
    public interface Natives {
        /**
         * Calls to MemoryMetricsLogger to force recording histograms, returning true on success.
         * A value of false means recording failed (most likely because process metrics not
         * available.
         */
        boolean forceRecordHistograms();
    }
}
